Gems/Atom/Tools/ShaderManagementConsole/Code/Source/Document/ShaderManagementConsoleDocument.cpp

```diff
@@ -215,15 +206,61 @@ namespace ShaderManagementConsole
 
     bool ShaderManagementConsoleDocument::IsModified() const
     {
-        return false;
+        return m_modified;
+    }
+
+    bool ShaderManagementConsoleDocument::BeginEdit()
+    {
+        // Save the current properties as a momento for undo before any changes are applied
+        m_shaderVariantListSourceDataBeforeEdit = m_shaderVariantListSourceData;
+        return true;
+    }
+
+    bool ShaderManagementConsoleDocument::EndEdit()
+    {
+        bool modified = false;
+
+        // Lazy evaluation, comparing the current and previous shader variant list source data state to determine if we need to record undo/redo history.
+        // TODO Refine this so that only the deltas are stored.
+        const auto& undoState = m_shaderVariantListSourceDataBeforeEdit;
+        const auto& redoState = m_shaderVariantListSourceData;
+        if (undoState.m_shaderFilePath != redoState.m_shaderFilePath ||
+            undoState.m_shaderVariants.size() != redoState.m_shaderVariants.size())
+        {
+            modified = true;
+        }
+        else
+        {
+            for (size_t i = 0; i < redoState.m_shaderVariants.size(); ++i)
+            {
+                if (undoState.m_shaderVariants[i].m_stableId != redoState.m_shaderVariants[i].m_stableId ||
+                    undoState.m_shaderVariants[i].m_options != redoState.m_shaderVariants[i].m_options)
+                {
+                    modified = true;
+                    break;
+                }
+            }
+        }
+
+        if (modified)
+        {
+            AddUndoRedoHistory(
+                [this, undoState]() { SetShaderVariantListSourceData(undoState); },
+                [this, redoState]() { SetShaderVariantListSourceData(redoState); });
+        }
+
+        m_shaderVariantListSourceDataBeforeEdit = {};
+        return true;
     }
 
     void ShaderManagementConsoleDocument::Clear()
     {
         AtomToolsFramework::AtomToolsDocument::Clear();
 
         m_shaderVariantListSourceData = {};
+        m_shaderVariantListSourceDataBeforeEdit = {};
         m_shaderAsset = {};
+        m_modified = {};
```

Comment:
Nit: Isn't `m_modified` a boolean? `m_modified = false;` would be more readable. (repeated)